File: src/lib/i18n/languages.ts
export type UiLanguage = "en" | "ar" | "pt" | "ru";

export type LanguageOption = {
  code: UiLanguage;
  label: string;
  nativeLabel: string;
  greeting: string;
  rtl: boolean;
};

export const DEFAULT_LANGUAGE: UiLanguage = "en";

export const LANGUAGES: LanguageOption[] = [
  { code: "en", label: "English", nativeLabel: "English", greeting: "Hello", rtl: false },
  { code: "ar", label: "Arabic", nativeLabel: "العربية", greeting: "مرحبا", rtl: true },
  { code: "pt", label: "Portuguese", nativeLabel: "Português", greeting: "Olá", rtl: false },
  { code: "ru", label: "Russian", nativeLabel: "Русский", greeting: "Привет", rtl: false },
];

const BY_CODE = new Map<string, LanguageOption>(LANGUAGES.map((l) => [l.code, l]));

export function normalizeLanguage(lang: unknown): UiLanguage {
  return typeof lang === "string" && BY_CODE.has(lang) ? (lang as UiLanguage) : DEFAULT_LANGUAGE;
}

export function isRtl(lang: UiLanguage): boolean {
  return BY_CODE.get(lang)?.rtl ?? false;
}

File: clients/lib/domain/i18n/languages.dart
/// The supported UI languages. Ported from `UiLanguage`/`LANGUAGES` in
/// `src/lib/i18n/languages.ts`. English is the source language (and the fallback
/// for any untranslated key).
class LanguageOption {
  const LanguageOption({
    required this.code,
    required this.label,
    required this.nativeLabel,
    required this.rtl,
  });

  final String code;
  final String label;
  final String nativeLabel;
  final bool rtl;
}

const languages = <LanguageOption>[
  LanguageOption(
    code: 'en',
    label: 'English',
    nativeLabel: 'English',
    rtl: false,
  ),
  LanguageOption(
    code: 'ar',
    label: 'Arabic',
    nativeLabel: 'العربية',
    rtl: true,
  ),
  LanguageOption(
    code: 'pt',
    label: 'Portuguese',
    nativeLabel: 'Português',
    rtl: false,
  ),
];

/// Whether [lang] lays out right-to-left. Ported from `isRtl`.
bool isRtl(String lang) => lang == 'ar';

/// Normalises an arbitrary setting value to a supported language code, else
/// English.
String normalizeUiLanguage(String? value) =>
    (value == 'ar' || value == 'pt') ? value! : 'en';
